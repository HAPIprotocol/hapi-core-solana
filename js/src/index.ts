export * from "./state";
export * from "./constants";
export { u64 } from "./utils";
export * from "./reader-client";
export * from "./reporter-client";
export * from "./authority-client";
export * from "./instructions/instructions";
