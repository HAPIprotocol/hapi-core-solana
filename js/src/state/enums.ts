export enum HapiAccountType {
  Uninitialized,
  Community,
  Network,
  Reporter,
  Case,
  Address,
}

export enum ReporterType {
  Inactive,
  Tracer,
  Full,
  Authority,
}

export enum Category {
  // Tier 0
  /// Safe
  Safe,

  // Tier 1 - Low risk
  /// Wallet service - custodial or mixed wallets
  WalletService,

  /// Merchant service
  MerchantService,

  /// Mining pool
  MiningPool,

  /// Exchange (Low Risk) - Exchange with high KYC standards
  LowRiskExchange,

  // Tier 2 - Medium risk
  /// Exchange (Medium Risk)
  MediumRiskExchange,

  /// DeFi application
  DeFi,

  /// OTC Broker
  OTCBroker,

  /// Cryptocurrency ATM
  ATM,

  /// Gambling
  Gambling,

  // Tier 3 - High risk
  /// Illicit organization
  IllicitOrganization,

  /// Mixer
  Mixer,

  /// Darknet market or service
  DarknetService,

  /// Scam
  Scam,

  /// Ransomware
  Ransomware,

  /// Theft - stolen funds
  Theft,

  // Tier 4 - Severe risk
  /// Terrorist financing
  TerroristFinancing,

  /// Sanctions
  Sanctions,

  /// Child abuse and porn materials
  ChildAbuse,
}
