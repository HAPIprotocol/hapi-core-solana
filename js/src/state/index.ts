export * from "./address";
export * from "./case";
export * from "./community";
export * from "./enums";
export * from "./network";
export * from "./reporter";
