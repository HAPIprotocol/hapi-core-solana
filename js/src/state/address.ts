import { Numberu64, Numberu8 } from "../utils";
import { Category, HapiAccountType } from "./enums";

export class Address {
  /// HAPI account type
  account_type: HapiAccountType;

  /// Risk score
  risk: Numberu8;

  /// Case ID
  case_id: Numberu64;

  /// Category
  category: Category;
}
