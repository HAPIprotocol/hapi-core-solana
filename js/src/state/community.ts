import { Connection, PublicKey } from "@solana/web3.js";
import { serialize, deserializeUnchecked } from "borsh";
import { u64 } from "utils";

import { HAPI_PROGRAM_ID } from "../constants";
import { HapiAccountType } from "./enums";

class CommunityState {
  account_type: number;
  authority: Uint8Array;
  name: string;
  next_case_id: u64;
  constructor(object: Partial<CommunityState>) {
    Object.assign(this, object);
  }
  static schema = new Map([
    [
      CommunityState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["authority", [32]],
          ["next_case_id", "u64"],
          ["name", "string"],
        ],
      },
    ],
  ]);
  static size = 73;
}

export class Community {
  /// HAPI account type
  accountType: HapiAccountType;

  /// HAPI authority account
  authority: PublicKey;

  /// ID of the next created case
  nextCaseId: u64;

  /// HAPI community name
  name: string;

  static size = CommunityState.size;

  constructor(data?: Partial<Community>) {
    if (data) {
      Object.assign(this, data);
    }
  }

  static async getAddress(communityName: string): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [Buffer.from("community"), Buffer.from(communityName)],
      HAPI_PROGRAM_ID
    );
  }

  static fromState(state: CommunityState): Community {
    return new Community({
      accountType: state.account_type,
      authority: new PublicKey(state.authority),
      nextCaseId: state.next_case_id,
      name: state.name,
    });
  }

  static deserialize(buffer: Buffer): Community {
    return Community.fromState(
      deserializeUnchecked(CommunityState.schema, CommunityState, buffer)
    );
  }

  static async retrieve(
    connection: Connection,
    communityName: string
  ): Promise<{ data: Community; account: PublicKey }> {
    const [communityAddress] = await Community.getAddress(communityName);

    const communityAccount = await connection.getAccountInfo(
      communityAddress,
      "confirmed"
    );
    if (!communityAccount) {
      throw new Error(
        `Community not found: ${communityName} (${communityAddress})`
      );
    }

    return {
      data: Community.deserialize(communityAccount.data),
      account: communityAddress,
    };
  }

  serialize(): Uint8Array {
    const buf = Buffer.alloc(CommunityState.size);
    buf.set(serialize(CommunityState.schema, this.toState()));
    return buf;
  }

  toState(): CommunityState {
    return new CommunityState({
      account_type: this.accountType,
      authority: this.authority.toBytes(),
      next_case_id: this.nextCaseId,
      name: this.name,
    });
  }
}
