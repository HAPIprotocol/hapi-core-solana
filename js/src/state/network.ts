import { Connection, PublicKey } from "@solana/web3.js";
import { deserializeUnchecked, Schema, serialize } from "borsh";

import { HAPI_PROGRAM_ID } from "../constants";
import { Community } from "./community";
import { HapiAccountType } from "./enums";

class NetworkState {
  account_type: number;
  name: string;
  constructor(object: Partial<NetworkState>) {
    Object.assign(this, object);
  }
  static schema: Schema = new Map([
    [
      NetworkState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["name", "string"],
        ],
      },
    ],
  ]);
  static size = 33;
}

export class Network {
  /// HAPI account type
  accountType: HapiAccountType;

  /// HAPI network name
  name: string;

  constructor(object?: Partial<Network>) {
    if (object) {
      Object.assign(this, object);
    }
  }

  static getAddress(
    communityAddress: PublicKey,
    networkName: string
  ): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [
        Buffer.from("network"),
        communityAddress.toBuffer(),
        Buffer.from(networkName),
      ],
      HAPI_PROGRAM_ID
    );
  }

  static fromState(state: NetworkState): Network {
    const network = new Network();
    network.accountType = state.account_type;
    network.name = state.name;
    return network;
  }

  static deserialize(buffer: Buffer): Network {
    return Network.fromState(
      deserializeUnchecked(NetworkState.schema, NetworkState, buffer)
    );
  }

  static async retrieve(
    connection: Connection,
    communityName: string,
    networkName: string
  ): Promise<{ data: Network; account: PublicKey }> {
    const [communityAddress] = await Community.getAddress(communityName);

    const [networkAddress] = await Network.getAddress(
      communityAddress,
      networkName
    );

    const account = await connection.getAccountInfo(
      networkAddress,
      "processed"
    );
    if (!account) {
      throw new Error(
        `Network not found: "${networkName}" (${networkAddress}) in community "${communityName}" (${communityAddress})`
      );
    }

    return { data: Network.deserialize(account.data), account: networkAddress };
  }

  serialize(): Uint8Array {
    const buf = Buffer.alloc(NetworkState.size);
    buf.set(serialize(NetworkState.schema, this.toState()));
    return buf;
  }

  toState(): NetworkState {
    return new NetworkState({
      account_type: this.accountType,
      name: this.name,
    });
  }
}
