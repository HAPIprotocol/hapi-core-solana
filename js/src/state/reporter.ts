import { Connection, PublicKey } from "@solana/web3.js";
import { deserialize, serialize } from "borsh";

import { HAPI_PROGRAM_ID } from "../constants";
import { Community } from "./community";
import { HapiAccountType, ReporterType } from "./enums";

class ReporterState {
  account_type: number;
  reporter_type: number;
  name: string;
  constructor(object: Partial<ReporterState>) {
    Object.assign(this, object);
  }
  static schema = new Map([
    [
      ReporterState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["reporter_type", "u8"],
          ["name", "string"],
        ],
      },
    ],
  ]);
}

export class Reporter {
  /// HAPI account type
  accountType: HapiAccountType;

  /// Reporter type
  reporterType: ReporterType;

  /// Reporter name
  name: string;

  constructor(data?: Partial<Reporter>) {
    if (data) {
      Object.assign(this, data);
    }
  }

  static async getAddress(
    communityAddress: PublicKey,
    reporterPubkey: PublicKey
  ): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [
        Buffer.from("reporter"),
        communityAddress.toBuffer(),
        reporterPubkey.toBuffer(),
      ],
      HAPI_PROGRAM_ID
    );
  }

  static fromState(state: ReporterState): Reporter {
    return new Reporter({
      accountType: state.account_type,
      reporterType: state.reporter_type,
      name: state.name,
    });
  }

  static deserialize(buffer: Buffer): Reporter {
    return Reporter.fromState(
      deserialize(ReporterState.schema, ReporterState, buffer)
    );
  }

  static async retrieve(
    connection: Connection,
    communityName: string,
    reporterPubkey: PublicKey
  ): Promise<Reporter> {
    const [communityAddress] = await Community.getAddress(communityName);

    const [reporterAddress] = await Reporter.getAddress(
      communityAddress,
      reporterPubkey
    );

    const account = await connection.getAccountInfo(
      reporterAddress,
      "processed"
    );
    if (!account) {
      throw new Error("Invalid reporter account provided");
    }

    return Reporter.deserialize(account.data);
  }

  serialize(): Uint8Array {
    return serialize(ReporterState.schema, this.toState());
  }

  toState(): ReporterState {
    return new ReporterState({
      account_type: this.accountType,
      reporter_type: this.reporterType,
      name: this.name,
    });
  }
}
