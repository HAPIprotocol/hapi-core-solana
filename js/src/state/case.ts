import { Connection, PublicKey } from "@solana/web3.js";
import { deserialize, serialize } from "borsh";

import { HAPI_PROGRAM_ID } from "../constants";
import { mapToBuffer, u64 } from "../utils";
import { Community } from "./community";
import { Categories, Category, HapiAccountType } from "./enums";
import { Network } from "./network";

class CaseState {
  account_type: number;
  name: string;
  reporter_key: Uint8Array;
  categories: Uint8Array;
  constructor(object: Partial<CaseState>) {
    Object.assign(this, object);
  }
  static schema = new Map([
    [
      CaseState,
      {
        kind: "struct",
        fields: [
          ["account_type", "u8"],
          ["name", "string"],
          ["reporter_key", [32]],
          ["categories", "u32"],
        ],
      },
    ],
  ]);
}

export class Case {
  /// HAPI account type
  accountType: HapiAccountType;

  /// Case name
  name: string;

  /// Case reporter key
  reporterKey: PublicKey;

  /// Categories
  categories: Map<Category, boolean>;

  constructor(data?: Partial<Case>) {
    if (data) {
      Object.assign(this, data);
      for (const category of Categories) {
        if (!this.categories.has(category)) {
          this.categories.set(category, false);
        }
      }
    }
  }

  static async getAddress(
    networkAddress: PublicKey,
    caseId: u64
  ): Promise<[PublicKey, number]> {
    return PublicKey.findProgramAddress(
      [Buffer.from("case"), networkAddress.toBuffer(), caseId.toBuffer()],
      HAPI_PROGRAM_ID
    );
  }

  static fromState(state: CaseState): Case {
    return new Case({
      accountType: state.account_type,
      name: state.name,
      reporterKey: new PublicKey(state.reporter_key),
      categories: new Map<Category, boolean>(),
    });
  }

  static deserialize(buffer: Buffer): Case {
    return Case.fromState(deserialize(CaseState.schema, CaseState, buffer));
  }

  static async retrieve(
    connection: Connection,
    communityName: string,
    networkName: string,
    caseId: u64
  ): Promise<Case> {
    const [communityAddress] = await Community.getAddress(communityName);

    const [networkAddress] = await Network.getAddress(
      communityAddress,
      networkName
    );

    const [caseAddress] = await Case.getAddress(networkAddress, caseId);

    const account = await connection.getAccountInfo(caseAddress, "processed");
    if (!account) {
      throw new Error("Invalid case account provided");
    }

    return Case.deserialize(account.data);
  }

  serialize(): Uint8Array {
    return serialize(CaseState.schema, this.toState());
  }

  toState(): CaseState {
    return new CaseState({
      account_type: this.accountType,
      name: this.name,
      reporter_key: this.reporterKey.toBytes(),
      categories: mapToBuffer(this.categories),
    });
  }
}
