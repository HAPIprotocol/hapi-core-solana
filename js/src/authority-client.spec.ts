jest.mock("rpc-websockets"); // disable web3's websocket connections

import { Keypair } from "@solana/web3.js";
import stringify from "fast-json-stable-stringify";
import nock from "nock";

import {
  mockConfirmTransaction,
  mockRpcError,
  mockRpcOk,
} from "../test/util/mocks";
import { HapiAccountType } from "./state";
import { AuthorityClient } from ".";

describe("AuthorityClient", () => {
  let client: AuthorityClient;
  let payer: Keypair;

  const endpoint = "http://localhost:8899";

  let i = 0;
  const seed = Buffer.alloc(32);
  beforeEach(async () => {
    nock.disableNetConnect();
    seed.writeUInt32BE(i++);
    payer = Keypair.fromSeed(seed);
    client = new AuthorityClient({ endpoint, payer });
  });

  afterEach(() => {
    nock.cleanAll();
    jest.clearAllMocks();
  });

  it("should initialize", () => {
    expect(client).toBeDefined();
  });

  describe("createCommunity", () => {
    it("should throw - community already exists", async () => {
      jest.spyOn(Date, "now").mockReturnValue(1633624165684);

      mockRpcOk(endpoint, "getRecentBlockhash", [], {
        context: { slot: 159 },
        value: {
          blockhash: "FMjzqMLtuECN4cKNZcu221QCBVExhJZqrKACi7pmw4qL",
          feeCalculator: { lamportsPerSignature: 5000 },
        },
      });

      mockRpcError(endpoint, "sendTransaction", [], {
        code: -32002,
        message:
          "Transaction simulation failed: Error processing Instruction 0: instruction requires an uninitialized account",
        data: {
          accounts: null,
          err: { InstructionError: [0, "AccountAlreadyInitialized"] },
          logs: [
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 invoke [1]",
            'Program log: HAPI-INSTRUCTION: CreateCommunity { name: "hapi.one" }',
            "Program log: Error: AccountAlreadyInitialized",
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 consumed 5872 of 200000 compute units",
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 failed: instruction requires an uninitialized account",
          ],
        },
      });

      jest.spyOn(console, "error").mockImplementationOnce(() => null);
      await expect(() =>
        client.createCommunity("hapi.one")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should throw - invalid name", async () => {
      await expect(() =>
        client.createCommunity("loooooooooooooooooooooooooooooooooong")
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it("should create a community - success", async () => {
      let t0 = 1633624165684;
      jest.spyOn(Date, "now").mockImplementation(() => (t0 += 1));

      mockConfirmTransaction(client);

      mockRpcOk(endpoint, "getRecentBlockhash", [], {
        context: { slot: 159 },
        value: {
          blockhash: "Gk8Vxi2AjCDBtAufgVKcnHv327sxEuTcZKUrNYusKhd4",
          feeCalculator: { lamportsPerSignature: 5000 },
        },
      });

      mockRpcOk(
        endpoint,
        "sendTransaction",
        [],
        "64cBhBwynYdXc6Ybr1suHM5Hz4xKtEzB3uyyg3gRtE8hfenzQJzd5whCZdPF5Xf5Quffvtib7U82XqYRQet4roU2"
      );

      mockRpcOk(
        endpoint,
        "getAccountInfo",
        ["Bwv5tFYijy58tNWKDv64Rs2i1Wun5C6LbxQEL2N35J59"],
        {
          context: { slot: 33 },
          value: {
            data: [
              "AatWllUf4XEcPDFmn/IOHgvBLLmZF8OrJBLnwTAT3ufnAAAAAAAAAAAJAAAAaGFwaS50ZXN0AAAAAAAAAAAAAAAAAAAAAAAAAA==",
              "base64",
            ],
            executable: false,
            lamports: 1398960,
            owner: "hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7",
            rentEpoch: 0,
          },
        }
      );

      const { account, data } = await client.createCommunity("hapi.test");
      expect(account.toString()).toEqual(
        "Bwv5tFYijy58tNWKDv64Rs2i1Wun5C6LbxQEL2N35J59"
      );
      expect(stringify(data)).toEqual(
        stringify({
          accountType: HapiAccountType.Community,
          authority: payer.publicKey,
          nextCaseId: "00",
          name: "hapi.test",
        })
      );
    });
  });

  describe("updateCommunity", () => {
    it.todo("should throw - community not found");
    it.todo("should update a community - success");
  });

  describe("createNetwork", () => {
    it("should throw - community not found", async () => {
      mockRpcOk(endpoint, "getRecentBlockhash", [], {
        context: { slot: 159 },
        value: {
          blockhash: "FMjzqMLtuECN4cKNZcu221QCBVExhJZqrKACi7pmw4qL",
          feeCalculator: { lamportsPerSignature: 5000 },
        },
      });

      mockRpcError(endpoint, "sendTransaction", [], {
        code: -32002,
        message:
          "Transaction simulation failed: Error processing Instruction 0: instruction requires an initialized account",
        data: {
          accounts: null,
          err: { InstructionError: [0, "UninitializedAccount"] },
          logs: [
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 invoke [1]",
            'Program log: HAPI-INSTRUCTION: CreateNetwork { name: "testcoin" }',
            "Program log: Error: UninitializedAccount",
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 consumed 6155 of 200000 compute units",
            "Program hapiScWyxeZy36fqXD5CcRUYFCUdid26jXaakAtcdZ7 failed: instruction requires an initialized account",
          ],
        },
      });

      let errorStack = "";
      jest.spyOn(console, "error").mockImplementationOnce((error, stack) => {
        errorStack = stack;
      });

      await expect(() =>
        client.createNetwork("testcoin", "community404")
      ).rejects.toThrowErrorMatchingSnapshot();

      expect(errorStack).toContain("Error: UninitializedAccount");
    });

    it.todo("should throw - network already exists");
    it.todo("should throw - invalid name");
    it.todo("should create a network - success");
  });

  describe("createReporter", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - reporter already exists");
    it.todo("should throw - invalid type");
    it.todo("should throw - invalid name");
    it.todo("should throw - invalid public key");
    it.todo("should create a reporter - success");
  });

  describe("updateReporter", () => {
    it.todo("should throw - community not found");
    it.todo("should throw - reporter not found");
    it.todo("should throw - invalid type");
    it.todo("should throw - invalid name");
    it.todo("should update a reporter - success");
  });
});
