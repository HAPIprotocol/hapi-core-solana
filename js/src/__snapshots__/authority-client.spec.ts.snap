// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorityClient createCommunity should throw - community already exists 1`] = `"failed to send transaction: Transaction simulation failed: Error processing Instruction 0: instruction requires an uninitialized account"`;

exports[`AuthorityClient createCommunity should throw - invalid name 1`] = `"Max seed length exceeded"`;

exports[`AuthorityClient createNetwork should throw - community not found 1`] = `"failed to send transaction: Transaction simulation failed: Error processing Instruction 0: instruction requires an initialized account"`;
